BUG: IntervalIndex.get_loc/get_indexer wrong return value / error

samuelsinayoko opened the pull request:
- [x] closes #25087
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

samuelsinayoko committed:
Revert earlier change and use to_numpy

Revert #24048 change that caused bug.

samuelsinayoko committed:
Remove warning by using to_numpy

samuelsinayoko committed:
Make warning go away

samuelsinayoko committed:
revert initial bugfix

Write tests first

samuelsinayoko committed:
Add test for contains in interval index categorical

When tested with a variable that has the wrong dtype, this raises an
exception instead of False

samuelsinayoko committed:
Check get_loc on interval index raises KeyError

When supplied a variable with the wrong type get_loc should raise a
KeyError (not type error). Otherwise things like checking if a variable
is in an index will fail.

samuelsinayoko committed:
Add test for get_indexer

samuelsinayoko committed:
Add a test for get_indexer with different type

samuelsinayoko committed:
Make first two tests pass

samuelsinayoko committed:
Make third test pass

This is enough for making the test pass but it's not the right implementation

rs2 commented on pandas/core/frame.py:
```diff
@@ -4615,6 +4615,7 @@ def dropna(self, axis=0, how='any', thresh=None, subset=None,
                 else:
                     raise TypeError('must specify how or thresh')
 
+            #result = self._take(mask.to_numpy().nonzero()[0], axis=axis)
```
This is the pre-existing implementation that can be located in git history. It's not needed here.

rs2 commented on pandas/core/indexes/interval.py:
```diff
@@ -766,8 +766,13 @@ def get_loc(self, key, method=None):
+            try:
+                start, stop = self._find_non_overlapping_monotonic_bounds(key)
+            except TypeError:
+                # get loc should raise KeyError
```
Typo: `.get_loc()`.

rs2 commented on pandas/core/indexes/interval.py:
```diff
@@ -766,8 +766,13 @@ def get_loc(self, key, method=None):
+                # get loc should raise KeyError
+                # if key is hashable but
+                # of an incorrect type
+                raise KeyError
```
This needs to be `raise from`, which is a Python 3 construct. To stay version-agnostic use `six.raise_from`.

```suggestion
import six
...
            try:
                start, stop = self._find_non_overlapping_monotonic_bounds(key)
            except TypeError as exc:
                six.raise_from(KeyError('Key is hashable, but of an incorrect type'), exc)
```

samuelsinayoko replied:
OK, I've left the code as is but have improved the reported message as suggested by @rs2

TomAugspurger replied:
We don't have a dependency on six, so we can't use `six.raise_from`.

I'd recommend just dropping the original exception from the traceback and leaving `TODO(py3): use raise from`.

rs2 commented on pandas/core/indexes/interval.py:
```diff
@@ -819,7 +824,14 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
+                )
+            except TypeError:
+                # This is probably wrong
+                # but not sure what I should do here
```
`¯\_(ツ)_/¯`

rs2 commented on pandas/core/indexes/interval.py:
```diff
@@ -819,7 +824,14 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
+            except TypeError:
+                # This is probably wrong
+                # but not sure what I should do here
+                return np.array([-1])
```
Please comment on the choice of `-1`.

rs2 commented:
xref: https://github.com/pandas-dev/pandas/pull/25091, https://github.com/pandas-dev/pandas/issues/25087

jschendel commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -886,6 +886,13 @@ def test_symmetric_difference(self, closed, sort):
+        index = pd.interval_range(0, 1)
+        # behaviour should be the same as Int64Index and return an
+        # array with values of -1
+        assert np.all(index.get_indexer(['gg']) == np.array([-1]))
```
use `tm.assert_numpy_array_equal` and make sure your expected is `'intp'` dtype.

jschendel commented on pandas/tests/indexing/test_loc.py:
```diff
@@ -778,3 +778,9 @@ def test_loc_setitem_empty_append_raises(self):
+        """ GH25087, test get_loc returns key error for interval indexes"""
+        idx = pd.interval_range(0, 1.0)
+        with pytest.raises(KeyError):
+            idx.get_loc('gg')
```
Instead of testing here, can you add this as a test case to `test_get_loc_value` in `pandas/tests/indexes/interval/test_interval.py`:

https://github.com/pandas-dev/pandas/blob/a09a07e4dabe91a8cfe46bec40414a2ddc4225a3/pandas/tests/indexes/interval/test_interval.py#L416

samuelsinayoko replied:
I've just implemented this in 93f75ea

jschendel commented on pandas/core/indexes/interval.py:
```diff
@@ -766,8 +766,13 @@ def get_loc(self, key, method=None):
                 key = self._maybe_cast_slice_bound(key, 'left', None)
-
-            start, stop = self._find_non_overlapping_monotonic_bounds(key)
+            try:
```
You'll also need to do something similar in the `else` branch to cover the overlapping/non-monotonic case, e.g. I think something like `pd.IntervalIndex.from_tuples([(1, 3), (2, 4), (0, 2)]).get_loc('foo')` will still fail.

jorisvandenbossche replied:
Maybe it should be the `engine` that should properly raise a KeyError? (eg the int64 engine does that)

samuelsinayoko replied:
I still need to look into @jorisvandenbossche's comment on raising the error in the engine itself (especially if that's the behaviour for int64), but I think I've addressed everything else.

jschendel replied:
@jorisvandenbossche : There is code in place within the `engine` that raises a `KeyError`, but strings queries fail before it gets there since the engine is expecting a `scalar_t` type (fused type consisting of numeric types) for `key`:
https://github.com/pandas-dev/pandas/blob/2e38d5552a5c7b2c0091cecddd483f4f08ad1d2c/pandas/_libs/intervaltree.pxi.in#L104

I'm not super well versed in Cython.  Is there a graceful way to force this to raise a `KeyError` within the Cython code?  Removing the `scalar_t` type gets a step further but still raises a `TypeError` as the code expects things to be comparable (probably some perf implications to removing it too).

jorisvandenbossche replied:
Yeah, the other engines have the key as `object` typed, and then afterwards do a check of that. 
But for me fine as well to leave that for now, and do the check here in the level above that. But on the long term would still be good to make the behaviour consistent throughout the different engines.

samuelsinayoko committed:
remove commented out code

samuelsinayoko committed:
Improve error message

samuelsinayoko committed:
Rename, move and parametrize indexer test

Include some non-monotonic/overlapping IntervalIndex.
This triggers another bug, due to the fact that self.get_loc(i)
is called on an unexpected key.

samuelsinayoko committed:
Use numpy_array_equal in indexer test

jorisvandenbossche commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -886,6 +886,13 @@ def test_symmetric_difference(self, closed, sort):
         result = index.symmetric_difference(other, sort=sort)
         tm.assert_index_equal(result, expected)
 
+    def test_interval_range_get_indexer_with_different_input_type(self):
```
I think you can ignore this one, already types the comment before saw the updates

samuelsinayoko committed:
Refactor interval index get_loc test

Move the test from indexing/test_loc to index/interval/test_interval.

samuelsinayoko committed:
Fix bug introduced in earlier commit

target was missing from call to _find_non_overlapping_monotonic_bounds

jorisvandenbossche commented on pandas/core/indexes/interval.py:
```diff
@@ -766,8 +766,11 @@ def get_loc(self, key, method=None):
+            try:
+                start, stop = self._find_non_overlapping_monotonic_bounds(key)
+            except TypeError:
+                # get_loc should raise KeyError
```
Can you add here a comment as Tom proposed (`TODO(py3): use raise from.`) ?

jreback replied:
you can now use raise from (as we are PY3 only)

jorisvandenbossche commented on pandas/core/indexes/interval.py:
```diff
@@ -766,8 +766,11 @@ def get_loc(self, key, method=None):
+                start, stop = self._find_non_overlapping_monotonic_bounds(key)
+            except TypeError:
+                # get_loc should raise KeyError
+                raise KeyError('key is hashable but of incorrect type')
```
I think for key errors we typically just pass the key itself as message, so might be good to be consistent with that. Or at least I would include the key in the message, something like: `"Key {0} is of the incorrect type".format(key)`

jorisvandenbossche commented on pandas/core/indexes/interval.py:
```diff
@@ -785,7 +788,10 @@ def get_loc(self, key, method=None):
+                try:
+                    return self._engine.get_loc(key)
+                except TypeError:
+                    raise KeyError('No engine for key {!r}'.format(key))
```
No need to mention "engine" here (that is something internal to pandas, while this error message will be visible for users)

jorisvandenbossche commented on pandas/core/indexes/interval.py:
```diff
@@ -819,7 +825,12 @@ def get_indexer(self, target, method=None, limit=None, tolerance=None):
+                    self._find_non_overlapping_monotonic_bounds(target)
+                )
+            except TypeError:
+                return np.repeat(np.int(-1), len(target))
```
@jschendel I am not fully sure this will cover all cases (but what was there before also not). 

`target` can in principle a mixture of valid and invalid keys. So maybe the easiest would be to fall back to the `else` branch that iterates through the elements separately in case this raises a TypeError.
That could be done by putting a `pass` here, and putting the `return` value of three lines below in the else part of a `try/except/else`.

samuelsinayoko replied:
Sorry getting a bit confused here. You're saying that we should pass on line 833 here and left the code run to the else branch starting on line 851 (non IntervalIndex) where it loops over each element and appends -1 to the list if a KeyError is raised (I would probably add TypeError too).

jorisvandenbossche replied:
Yes, that is basically what I meant I think.

So that if you do `idx.get_indexer(['a', 1])` (where 1 is a valid key), you get [-0, 1] as result instead of [-1, -1] (assuming that `idx.get_loc(1)` would return 1)

samuelsinayoko replied:
OK I've just pushed a commit implementing this. I had to make a few tweaks to the "non IntervalIndex" else branch (starting line 849 in core/indexes/interval.py) to make my two new tests pass, but hopefully haven't broken anything. See 2c48272

jorisvandenbossche commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -616,6 +624,14 @@ def test_get_indexer_length_one(self, item, closed):
         tm.assert_numpy_array_equal(result, expected)
 
+    @pytest.mark.parametrize('index', [
+        pd.interval_range(0, 1),
```
I would add here one that is a bit larger as well (like `pd.interval_range(0, 3)`), as 1-element indices can behave a bit special regarding monotonicity / sortedness (as they always are that)

jorisvandenbossche commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -616,6 +624,14 @@ def test_get_indexer_length_one(self, item, closed):
+    ])
+    def test_get_indexer_errors(self, index):
+        expected = np.array([-1], dtype='intp')
+        assert tm.assert_numpy_array_equal(index.get_indexer(['gg']), expected)
```
Can you here also test multiple values and a mixture of values? Like `get_indexer(['a', 'b']` and `get_indexer([1, 'a'])`

jorisvandenbossche replied:
Can you do this one?

samuelsinayoko replied:
Sure will look into this over the weekend.

samuelsinayoko replied:
see 02127ff

jorisvandenbossche commented on pandas/tests/indexes/interval/test_interval.py:
```diff
@@ -435,6 +435,14 @@ def test_get_loc_value(self):
         idx = IntervalIndex.from_arrays([0, 2], [1, 3])
         pytest.raises(KeyError, idx.get_loc, 1.5)
 
+        # GH25087, test get_loc returns key error for interval indexes
```
Can you put this in a new test? (you can leave it in this place, but just put a `def test_get_loc_invalid_key(self)` above this line) 
Reason is that the other test is commented to be replaced, but this new test we want to keep.

jorisvandenbossche replied:
Can you do this one?

samuelsinayoko replied:
Sure.

samuelsinayoko replied:
Have got a few broken tests to fix. Hoping to make the build green in the coming days.

samuelsinayoko committed:
Add reminder comment to use raise from for python 3

samuelsinayoko committed:
Include key in error message.

samuelsinayoko committed:
Add larger interval range to test

samuelsinayoko committed:
get_loc should raise KeyError if the supplied key has the wrong type

samuelsinayoko committed:
Only return -1 in get_indexer for incorrect values

Instead of returning [-1, -1, -1] when the middle value is incorrect
type, return [a, -1, b].

samuelsinayoko committed:
Better tests for get_indexer_errors

Add mix of invalid and valid values

samuelsinayoko committed:
Fix broken test in test_interval

Fixes test_with_overlaps test

samuelsinayoko committed:
Fix broken tests in test_concat

interval.get_indexer() should still raise a TypeError in cases
where the types are unorderable. This is needed for DataFrame.append
for example, which was breaking tests in test_concat.

samuelsinayoko commented:
Any tips on making the tests pass on macOS, windows and Linux py27?

TomAugspurger commented:
@samuelsinayoko do those tests fail for you locally? If not, I can take a look later.

samuelsinayoko commented:
will take another look at this tomorrow

On Thu, 7 Mar 2019 at 18:00, Tom Augspurger <notifications@github.com>
wrote:

> @samuelsinayoko <https://github.com/samuelsinayoko> do those tests fail
> for you locally? If not, I can take a look later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/25090#issuecomment-470630333>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQuwvnphRf-gB8HLdg_PKKyjG_Yfqc15ks5vUVOkgaJpZM4afkZ_>
> .
>


-- 

*Sam Sinayoko* | Data Scientist

*BMLL Technologies Ltd.*

10th Floor – Portland House

Bressenden Place

London

SW1E 5RS

+44(0)20 3828 9020

samuelsinayoko commented:
@TomAugspurger yes the tests pass locally on my linux machine. Not sure what's happening on Windows and OSX, so would be grateful for any help on this.

WillAyd commented:
@samuelsinayoko can you merge master? Recently dropped Py2 support so should make things easier

samuelsinayoko commented:
Ok will do

samuelsinayoko committed:
Merge remote-tracking branch 'upstream/master' into 23264

jreback commented:
@samuelsinayoko can you merge mater and update

jreback commented:
closing as stale, if you'd like to continue, pls ping.
